pkg/downloader/manager.go

```diff
@@ -86,6 +86,11 @@ func (m *Manager) Build() error {
 	}
 
 	if sum, err := resolver.HashReq(req, lock.Dependencies); err != nil || sum != lock.Digest {
+		// If lock digest differs and chart is apiVersion v1, it maybe because the lock was built
+		// with Helm 2 and should therefore be regenerated
+		if c.Metadata.APIVersion == chart.APIVersionV1 {
+			return m.Update()
```

Comment:
I feel like Update is the wrong thing here, because that might pull in *different* semver-matching versions of packages, ignoring the versions *locked* in the Lockfile.

Can we detect an APIVersionV1 Hash and recognise it?

Reply:
I agree that update is a "destructive" operation, and thus should not be used here. 

i am wondering if we have to add a function on the Manager that re-calculates the hashes for v2. But even there, I am a little bit uncertain.

Reply:
How hard would it be to make `resolver.HasReq(req, nil)` generate the same hash as Helm v2? Did `chart.Dependencies` change a lot?

Reply:
The semver package is a major version change between Helm v2 and v3. Behind semver some rules were mildly modified in how they calculate versions. In order to calculate the rules the same as Helm v2 we would need to pull in semver v1 and perform the calculation using that. You can read about the changes [here](https://github.com/Masterminds/semver/blob/master/CHANGELOG.md#300-2019-09-12).

Reply:
@mattfarina Thanks for the details about the semver.

@technosophos @TBBle FYI, I went with the "nuclear" approach first as I could be certain about the result. I am not however happy with this approach (or even doing a hashing function to map from v2 to v3) as I feel we are circumventing a valid error being raised.

I think that @technosophos suggestion in https://github.com/helm/helm/issues/7233#issuecomment-579023372 about adding a warning with a specific message might be the best approach.

I am open to ideas on how best to work this.

Reply:
@mattfarina Those changes seem to be about how the semver is matched? The Helm v2 HashReq was only looking at the pre-match expression (as a string), so I don't think it's a major issue if we accept whatever was selected at `helm dep update` time, if the has tells us that the requirements.yaml (and for v3 requirements.lock) are the same as at that time, even if `helm dep up` *now* would pick different versions.

The structure itself (`Dependency`) is unchanged, unless hashing picks up the package of the type.

I don't think this is a valid failure to raise as an error, personally. The chart *is* valid under Helm v2, a newer version of Helm shouldn't suddenly consider the chart *invalid*. We actually have the converse problem that `APIVersion1` charts produced with Helm v3 would fail this check with Helm v2, except I think the check didn't exist in Helm v2?

Reply:
> The chart is valid under Helm v2, a newer version of Helm shouldn't suddenly consider the chart invalid.

@TBBle I appreciate your justification on why this is needed.

I have pushed changes which will now check against v2 style hash. let me know what you think.